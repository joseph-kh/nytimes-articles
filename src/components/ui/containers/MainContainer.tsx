import React, { useRef, useEffect } from "react";
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  Keyboard,
} from "react-native";
import { useHeaderHeight } from "@react-navigation/elements";

type Props = {
  isKeyboardAvoidingView?: boolean | undefined;
  children?: React.ReactNode;
};

type Ref = ScrollView;

const MainContainer = React.forwardRef<Ref, Props>((props, ref) => {
  const { children, isKeyboardAvoidingView } = props;

  const height = useHeaderHeight();

  const scrollRef = useRef<ScrollView | null>(null);

  useEffect(() => {
    if (Platform.OS === "ios" && isKeyboardAvoidingView) {
      const showSubscription = Keyboard.addListener("keyboardWillShow", () => {
        scrollRef.current?.scrollToEnd({ animated: true });
      });
      return () => showSubscription.remove();
    }
  }, []);

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === "ios" ? "padding" : undefined}
      style={{ flex: 1 }}
      keyboardVerticalOffset={height}
      enabled={isKeyboardAvoidingView}
    >
      <SafeAreaView style={styles.mainContainerStyle}>
        <ScrollView
          bounces={false}
          contentContainerStyle={{ flexGrow: 1 }}
          keyboardShouldPersistTaps="handled"
          ref={ref || scrollRef}
        >
          {children}
        </ScrollView>
      </SafeAreaView>
    </KeyboardAvoidingView>
  );
});

const styles = StyleSheet.create({
  mainContainerStyle: {
    flex: 1,
  },
});

export default MainContainer;
