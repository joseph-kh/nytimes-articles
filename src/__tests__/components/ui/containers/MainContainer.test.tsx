import React from "react";
import MainContainer from "@/components/ui/containers/MainContainer";
import { act, create } from "react-test-renderer";

jest.mock("@react-navigation/elements", () => ({
  useHeaderHeight: () => 30,
}));

describe("MainContainer Component", () => {
  it("renders correctly without KeyboardAvoidingView", async () => {
    await act(async () => {
      const tree = create(<MainContainer />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it("renders correctly with KeyboardAvoidingView", async () => {
    await act(async () => {
      const tree = create(
        <MainContainer isKeyboardAvoidingView={true} />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
