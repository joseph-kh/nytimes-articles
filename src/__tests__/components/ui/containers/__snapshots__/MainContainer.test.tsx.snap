// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContainer Component renders correctly with KeyboardAvoidingView 1`] = `null`;

exports[`MainContainer Component renders correctly without KeyboardAvoidingView 1`] = `null`;
